src/qibo/derivative.py

```diff
@@ -156,5 +233,281 @@ def circuit(nqubits = 1):
 
     # float() necessary to not return a 0-dim ndarray
     result = float(generator_eigenval * (forward - backward) * scale_factor)
+    return result
+
+
+def generate_new_stochastic_params(Param, ipar):
+    """Generates the three-gate parameters needed for the stochastic parameter-shift rule"""
+
+    sampling = random.random()
+    trainable_param = Param._variational_parameters[ipar]
+    F = Param.get_fixed_part(ipar)
+    scaling = Param.get_scaling_factor(ipar)
+
+    return [sampling * F, trainable_param, (1 - sampling) * F], scaling
+
+
+def stochastic_parameter_shift(
+    circuit,
+    hamiltonian,
+    parameter_index,
+    parameter: Parameter,
+    initial_state=None,
+    nshots=None,
+):
+    """In this method the stochastic parameter shift rule (SPSR) is implemented.
+    Given a circuit U and an observable H, the SPSR allows to calculate the derivative
+    of the expected value of H on the final state with respect to a variational
+    parameter of the circuit. The SPSR can calculate gradient approximations on
+    a larger family of gates than the standard PSR.
+    There is also the possibility of setting a scale factor. It is useful when a
+    circuit's parameter is obtained by combination of a variational
+    parameter and an external object, such as a training variable in a Quantum
+    Machine Learning problem. For example, performing a re-uploading strategy
+    to embed some data into a circuit, we apply to the quantum state rotations
+    whose angles are in the form: theta' = theta * x, where theta is a variational
+    parameter and x an input variable. The PSR allows to calculate the derivative
+    with respect of theta' but, if we want to optimize a system with respect its
+    variational parameters we need to "free" this procedure from the x depencency.
+    If the `scale_factor` is not provided, it is set equal to one and doesn't
+    affect the calculation.
```

Comment:
It seems to me you are not using the scale factor here. 
Consider to drop this part of the docstring or to pass the factor as argument.